### PR TITLE
A few test tweaks

### PR DESCRIPTION
**Pull request by ljedrz**
This PR contains the following changes:
- `ClientNode::number_of_connected_peers` is added (for test use)
- `snarkos_environment::TestEnvironment` is added (in order to have fewer max peers and an empty list of `SYNC_NODES` in tests, and possibly further tweaks for future tests)
- the `wait_until` macro is used in the `connection_limits_are_obeyed` test (to avoid flakiness in CI runs)
- the workspace members with the `test` feature now use it by default in tests
- new performance-oriented integration tests
- RPC test code deduplication
- faster `snarkos-storage` tests (93s > 41s on my PC)

note: I've considered feature-gating `TestEnvironment`, but there isn't much reason to do that, as it's basically free to compile

### CODE REVIEW

**Review by niklaslong — COMMENTED** [review]
LGTM 👍

**Review by niklaslong — APPROVED** [review]

**Comment by ljedrz** [comment]
In addition, a few more perf-oriented integration tests were added.

edit: added further test tweaks, and updated the PR description to contain all the current changes.

File: Cargo.toml [hunk 1]
```diff
@@ -67,6 +67,10 @@ features = [
   "compression-flate2"
 ]
 
+[dev-dependencies.snarkos]
+path = "."
+features = [ "test" ] # a workaround to use the `test` feature in tests by default
+
```
Comment (howardwu) [inline Cargo.toml]: nit: move this to the `dev-dep` section below the `dep` declarations

File: storage/src/state/tests.rs [hunk 2]
```diff
@@ -218,60 +196,21 @@ fn test_get_block_locators() {
...
-        ledger
-            .check_block_locators(&block_locators)
-            .expect("Failed to check block locators")
-    );
+    // Load test blocks.
+    let test_blocks = test_blocks_3();
+
+    // Check the block locators after each block insertion.
+    for block in test_blocks {
+        ledger.add_next_block(&block).expect("Failed to add next block to ledger");
```
Comment (howardwu) [inline storage/src/state/tests.rs:204]:
```suggestion
        // Mine the next block.
        ledger.add_next_block(&block).expect("Failed to add next block to ledger");
```

File: storage/src/state/tests.rs [hunk 3]
```diff
@@ -218,60 +196,21 @@ fn test_get_block_locators() {
...
+        let block_locators = ledger
```
Comment (howardwu) [inline storage/src/state/tests.rs:205]:
```suggestion
        // Check the block locators.
        let block_locators = ledger
```

File: storage/src/state/tests.rs [hunk 4]
```diff
@@ -22,12 +22,23 @@ use snarkos_environment::CurrentNetwork;
...
 
 fn temp_dir() -> std::path::PathBuf {
     tempfile::tempdir().expect("Failed to open temporary directory").into_path()
 }
 
+/// Returns 3 test blocks.
+fn test_blocks_3() -> Vec<Block<CurrentNetwork>> {
+    let mut test_block_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    test_block_path.push("benches");
+    // note: the `blocks_3` file was generated on a testnet2 storage using `LedgerState::dump_blocks`.
```
Comment (howardwu) [inline storage/src/state/tests.rs]: nit: Capitalize `note` to `Note`

Comment (howardwu) [reply storage/src/state/tests.rs]: nit: Move this to L32 (right before the `fn test_blocks_3` declaration)

**Comment by ljedrz** [comment]
Rebased and applied review suggestions.

**Review by howardwu — APPROVED** [review]

**Pull request merged**